### ISSUE TITLE
error when AnonymousClassMergingPass processing miranda method.

### ISSUE THREAD

**Issue by fengruisd** [facebook/redex#739]
it seems like AnonymousCM can not find the 'overridden method' if the top method of a `VirtualScope` is a miranda method, and then AnonymousCM can not merge the classes correctly. 

this is my case
```
public interface Intf {
  void method();
}
// abstract base, does not implement any method
public abstract class AbsBase implements Intf {
}
public class RealBase extends AbsBase {
  public void method() { print("method in RealBase"); }
}

public class Sub1 extends RealBase {
  public void method() { print("method in Sub1"); }
}
public class Sub2 extends RealBase {
}
```
Suppose Sub1 and Sub2 are anonymous classes and we want to invoke Sub2.method() in somewhere, the expected behavior is to print 'method in RealBase' and the class should be
```
public class GeneratedClassByAnonymousCM {
  // ...
  public void method() {
   switch ($t) {
    case 0:
      method(this); 
      break;
    case 1:
      super.method();   // should call super method for Sub2
    break;
   }
  }
  public static void method(GeneratedClassByAnonymousCM thisObj) {
    print("method in Sub1");
  }
}
```

but the generated class is as follows
```
public class GeneratedClassByAnonymousCM {
  // ...
  public void method() {
    method(this); 
  }
  public static void method(GeneratedClassByAnonymousCM thisObj) {
    print("method in Sub1");
  }
}
```
I debugged the code and find that `overridden_meth` is null, because the top method is a miranda method and `is_def()` return false here, link: [Model.cpp](https://github.com/facebook/redex/blob/main/service/class-merging/Model.cpp#L1182)

I tried to make `overridden_​​meth = top_def` when `top_def & miranda != 0` and it works in my case. But I'm not sure if it's suitable for a more general case.

**Comment by thezhangwei** [comment]
Thanks for reporting the issue. Yes it's on the todo list but we haven't had a chance to get to it. The main reason is I've not run into a case like this specially for anonymous classes. They in most cases simply overwrites a single virtual callback.

We are working on a fix.

**Comment by thezhangwei** [comment]
Hi, can you verify if 18cd5ff821cac7f4a3136fdbb7b6ba0bfe24f6b2 fixes the issue?

**Comment by fengruisd** [comment]
> Hi, can you verify if [18cd5ff](https://github.com/facebook/redex/commit/18cd5ff821cac7f4a3136fdbb7b6ba0bfe24f6b2) fixes the issue?

There is still a bad case in my project. 

`Animator.AnimatorListener` is an abstract class in AOSP and implements `AnimatorListener` which contains several default methods.

the methods in `Animator.AnimatorListener` are actually concrete methods but some of them (the methods which wrapped default methods) are not explored in `android.jar`, so redex mark them as miranda.

I am not sure if i make it clear so i write a demo to reproduce the problem.

```
package com.demo;

public class Demo {
    public static Animator.AnimatorListener createListener(int type) {
        if (type > 0) {
            return new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    System.out.println("$1");
                }
            };
        } else {
            return new AnimatorListenerAdapter() {

                @Override
                public void onAnimationEnd(Animator animation, boolean isReverse) {
                    System.out.println("$2");
                }
            };
        }
    }
}
```
the generated method does not invoke super method for default case
```
public class ALAdapterS0S0000000 extends AnimatorListenerAdapter {
    public void onAnimationEnd(Animator animator) {
        switch (this.$t) {
            case 0:
                onAnimationEnd(this, animator);
                return;
            default:
                super.onAnimationEnd(animator);
                return;
        }
    }

    public void onAnimationEnd(Animator animator, boolean z) {
        ///////////////////// does not  invoke super method for default case
        onAnimationEnd(this, animator, z); 
    }

    public static void onAnimationEnd(ALAdapterS0S0000000 aLAdapterS0S0000000, Animator animator) {
        System.out.println("$1");
    }
    public static void onAnimationEnd(ALAdapterS0S0000000 aLAdapterS0S0000000, Animator animator, boolean z) {
        System.out.println("$2");
    }
}
```
I uploaded the apk and redex.config ([apk_and_config.zip](https://github.com/facebook/redex/files/10001392/apk_and_config.zip)), and execute redex by command
```
redex --android-sdk-path /Users/user0/Android/sdk -o ./app-release-opt.apk -c ./redex.config ./app-release.apk
```

**Comment by fengruisd** [comment]
my android.jar level is 32, and i think this problem can be reproduced on most versions of android.jar.

i removed RebindRefs and the problem still exist

**Comment by thezhangwei** [comment]
I suspect the android.jar is not loaded properly. I created a similar test case, and it worked for me with the synthesized default methods.
Could you try: `-j ~/Users/user0/Android/sdk/platforms/android-32/android.jar` in addition to the `--android-sdk-path` option?

**Comment by fengruisd** [comment]
> I suspect the android.jar is not loaded properly. I created a similar test case, and it worked for me with the synthesized default methods. Could you try: `-j ~/Users/user0/Android/sdk/platforms/android-32/android.jar` in addition to the `--android-sdk-path` option?

I will try it. and can you attach your test case so i can refer to it

**Comment by thezhangwei** [comment]
Here is the test case:
	
```
interface IntfWithDefaults {
  default String methodOne() {
    return "IntWithDefaultsOne";
  }
  default String methodTwo(boolean f) {
    return "IntWithDefaultsTwo";
  }
}
	
class SubE1 implements IntfWithDefaults {
  public String methodOne() { return "SubE1"; }
}
class SubE2 implements IntfWithDefaults {
  public String methodTwo(boolean f) {
    return "SubE2";
  }
}
```
Here is the post dexdump:
```
    #0              : (in Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;)
      name          : 'methodOne'
      type          : '()Ljava/lang/String;'
      access        : 0x0001 (PUBLIC)
      code          -
      registers     : 2
      ins           : 1
      outs          : 1
      insns size    : 26 16-bit code units
080b28:                                        |[080b28] com.facebook.redex.MirandaWDefaultsShape4S0100000.methodOne:()Ljava/lang/String;
080b38: 5210 1105                              |0000: iget v0, v1, Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;.$t:I // field@0511
080b3c: 2b00 1000 0000                         |0002: packed-switch v0, 00000012 // +00000010
080b42: 1200                                   |0005: const/4 v0, #int 0 // #0
080b44: 1100                                   |0006: return-object v0
080b46: 7110 5509 0100                         |0007: invoke-static {v1}, Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;.methodOne:(Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;)Ljava/lang/String; // method@0955
080b4c: 0c00                                   |000a: move-result-object v0
080b4e: 28fb                                   |000b: goto 0006 // -0005
080b50: 7110 5609 0100                         |000c: invoke-static {v1}, Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;.methodOne$ClassMergingMirandaTest$SubE2:(Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;)Ljava/lang/String; // method@0956
080b56: 0c00                                   |000f: move-result-object v0
080b58: 28f6                                   |0010: goto 0006 // -000a
080b5a: 0000                                   |0011: nop // spacer
080b5c: 0001 0200 0000 0000 0500 0000 0a00 ... |0012: packed-switch-data (8 units)
      catches       : (none)
      positions     :
      locals        :

    #1              : (in Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;)
      name          : 'methodTwo'
      type          : '(Z)Ljava/lang/String;'
      access        : 0x1001 (PUBLIC SYNTHETIC)
      code          -
      registers     : 3
      ins           : 2
      outs          : 2
      insns size    : 32 16-bit code units
080bbc:                                        |[080bbc] com.facebook.redex.MirandaWDefaultsShape4S0100000.methodTwo:(Z)Ljava/lang/String;
080bcc: 5210 1105                              |0000: iget v0, v1, Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;.$t:I // field@0511
080bd0: 2b00 1600 0000                         |0002: packed-switch v0, 00000018 // +00000016
080bd6: 1200                                   |0005: const/4 v0, #int 0 // #0
080bd8: 1100                                   |0006: return-object v0
080bda: 0710                                   |0007: move-object v0, v1
080bdc: 1f00 2602                              |0008: check-cast v0, Lcom/facebook/redex/MirandaWDefaultsShape4S0100000; // type@0226
080be0: 7120 5709 2000                         |000a: invoke-static {v0, v2}, Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;.methodTwo:(Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;Z)Ljava/lang/String; // method@0957
080be6: 0c00                                   |000d: move-result-object v0
080be8: 28f8                                   |000e: goto 0006 // -0008
080bea: 0710                                   |000f: move-object v0, v1
080bec: 1f00 2602                              |0010: check-cast v0, Lcom/facebook/redex/MirandaWDefaultsShape4S0100000; // type@0226
080bf0: 7120 5909 2000                         |0012: invoke-static {v0, v2}, Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;.methodTwo$ClassMergingMirandaTest$SubE2:(Lcom/facebook/redex/MirandaWDefaultsShape4S0100000;Z)Ljava/lang/String; // method@0959
080bf6: 0c00                                   |0015: move-result-object v0
080bf8: 28f0                                   |0016: goto 0006 // -0010
080bfa: 0000                                   |0017: nop // spacer
080bfc: 0001 0200 0000 0000 0500 0000 0d00 ... |0018: packed-switch-data (8 units)
```

Of course, no external root is involved here. So it's not clear what hierarchy is actually loaded into Redex in your use case.
The dexer output is another factor. We use D8 as our dexer. If an implementer class does not override a default interface method, D8 will synthesize a virtual method calling the default implementation for it:
```
    #1              : (in Lcom/facebook/redextest/ClassMergingMirandaTest$SubE1;)
      name          : 'methodTwo'
      type          : '(Z)Ljava/lang/String;'
      access        : 0x1001 (PUBLIC SYNTHETIC)
      code          -
      registers     : 2
      ins           : 2
      outs          : 2
      insns size    : 5 16-bit code units
04ab54:                                        |[04ab54] com.facebook.redextest.ClassMergingMirandaTest$SubE1.methodTwo:(Z)Ljava/lang/String;
04ab64: 7120 7e09 1000                         |0000: invoke-static {v0, v1}, Lcom/facebook/redextest/IntfWithDefaults$-CC;.$default$methodTwo:(Lcom/facebook/redextest/IntfWithDefaults;Z)Ljava/lang/String; // method@097e
04ab6a: 0c01                                   |0003: move-result-object v1
04ab6c: 1101                                   |0004: return-object v1
```
In this case, Class Merging's job is pretty straightforward.

In the short term, maybe the simplest way to have you unblocked is to put the root in the blocklist of the pass. I'm heading out for a vacation and won't be very responsive for the rest of the year. :)

**Comment by thezhangwei** [comment]
Hey, just wanted to check in to see if the issue is resolved. Do you still need assistance from us?

**Comment by thezhangwei** [comment]
Pls open another issue if you have more follow ups. :)

**Comment by thezhangwei** [comment]
Hey @fengruisd, we recently landed a couple of fixes in Class Merging around external interface methods: https://github.com/facebook/redex/commit/706fe6cbd2ee4404e1e51e4653f7a9e21986203b and https://github.com/facebook/redex/commit/2373385d249ac48a153deb67f658c4fa4a1f4ffb. The 1st one fixes the fallback case for methods overriding external default methods. The 2nd one adds an additional check in `MergeabilityChecker` that rejects mergeable classes if they override external methods do not exist in the min_sdk API level of the app.

They are seemingly related to the issue you were reporting. So I just wanted to let you know that we landed new fixes, and hopefully they address your issue if it's not fixed already. Cheers,
